cns/restserver/ipam_test.go

```diff
@@ -0,0 +1,291 @@
+// Copyright 2020 Microsoft. All rights reserved.
+// MIT License
+
+package restserver
+
+import (
+	"encoding/json"
+	"reflect"
+	"testing"
+
+	"github.com/Azure/azure-container-networking/cns"
+	"github.com/Azure/azure-container-networking/cns/common"
+)
+
+var (
+	testNCID = "06867cf3-332d-409d-8819-ed70d2c116b0"
+
+	testIP1      = "10.0.0.1"
+	testPod1GUID = "898fb8f1-f93e-4c96-9c31-6b89098949a3"
+	testPod1Info = cns.KubernetesPodInfo{
+		PodName:      "testpod1",
+		PodNamespace: "testpod1namespace",
+	}
+
+	testIP2      = "10.0.0.2"
+	testPod2GUID = "b21e1ee1-fb7e-4e6d-8c68-22ee5049944e"
+	testPod2Info = cns.KubernetesPodInfo{
+		PodName:      "testpod2",
+		PodNamespace: "testpod2namespace",
+	}
+
+	testPod3GUID = "718e04ac-5a13-4dce-84b3-040accaa9b41"
+	testPod3Info = cns.KubernetesPodInfo{
+		PodName:      "testpod3",
+		PodNamespace: "testpod3namespace",
+	}
+)
+
+func getTestService() *HTTPRestService {
+	var config common.ServiceConfig
+	httpsvc, _ := NewHTTPRestService(&config)
+	svc := httpsvc.(*HTTPRestService)
+	svc.state.OrchestratorType = cns.Kubernetes
+
+	return svc
+}
+
+// Want first IP
+func TestIPAMGetAvailableIPConfig(t *testing.T) {
+	svc := getTestService()
+
+	desiredState := NewPodState(testIP1, 24, testPod1GUID, testNCID, cns.Available)
+	ipconfigs := []*cns.ContainerIPConfigState{
+		desiredState,
+	}
+	svc.AddIPConfigsToState(ipconfigs)
+
+	req := cns.GetIPConfigRequest{}
+	b, _ := json.Marshal(testPod1Info)
+	req.OrchestratorContext = b
+
+	actualstate, err := requestIPConfigHelper(svc, req)
+	if err != nil {
+		t.Fatal("Expected IP retrieval to be nil")
+	}
+
+	desiredState.State = cns.Allocated
+	desiredState.OrchestratorContext = b
+	if reflect.DeepEqual(desiredState, actualstate) != true {
```

Comment:
I think the requestIPConfigHelper is also modifying the expected state. I think at line 63 desiredState.State has already been changed to cns.Allocated because requestIPConfigHelper modified it, which feels a bit weird.

The same applies for other tests in this file as well. I think you should do what line 98 does for all tests, that is to create a new desiredState object for comparison with actualState

Reply:
yep, sounds good, fixed.